File: 2015/Day_02/Ruby/part_2.rb
require "../../input_parsing/parse_input"

def get_dimensions(dimension_line)
    dimensions = dimension_line.split('x')
    dimensions.map(&:to_i)
end


def calculate_bow_length(dimensions)
    dimensions[0]*dimensions[1]*dimensions[2]
end

def calculate_ribbon_length(dimensions)
    sorted_dimensions = dimensions.sort
    2* sorted_dimensions[0] + 2 * sorted_dimensions[1]
end
    

if $PROGRAM_NAME == __FILE__
    input_text = input_per_line('../input.txt')
    all_ribbon_lengths = []
    all_bow_lengths = []
    input_text.each do |box|
        all_ribbon_lengths.append(calculate_ribbon_length(get_dimensions(box)))
        all_bow_lengths.append(calculate_bow_length(get_dimensions(box)))
    end

    summed_ribbon_lengths = all_ribbon_lengths.reduce(0) {|sum, num| sum + num}
    summed_bow_lengths = all_bow_lengths.reduce(0) {|sum, num| sum + num}

    puts "The Elves need #{summed_ribbon_lengths+summed_bow_lengths} feet of ribbon!"
end

File: 2015/Day_02/Python/part_2.py
import sys
sys.path.insert(0, '../../input_parsing')
from functools import reduce
import parse_input


def get_dimensions(dimensions: str):
    """Take in a string of LxWxD where L, W, and D are numbers.

    :returns: A list of numbers.
    """
    dimensions_as_strings = dimensions.split('x')
    return [int(number) for number in dimensions_as_strings]


def determine_lengths(dimension):
    """Take a list with numbers and find out the length of ribbon and bow."""
    ribbon_length = sorted(dimension)[0] * 2 + sorted(dimension)[1] * 2
    bow_length = dimension[0] * dimension[1] * dimension[2]
    return ribbon_length, bow_length


def full_length(two_lengths):
    """Take the tuple with the ribbon and bow lengths and combine them."""
    return two_lengths[0] + two_lengths[1]


def add_up_all_boxes(box_dimension_list):
    """Take a list of areas and add them all up."""
    all_areas = [full_length(determine_lengths(get_dimensions(box))) for box in box_dimension_list]
    return reduce(lambda a, b: a + b, all_areas)


if __name__ == "__main__":
    box_dimensions = parse_input.input_per_line('../input.txt')
    print(f"The Elves need {add_up_all_boxes(box_dimensions)} feet of ribbon")
